### PR TITLE
Start of netcheck package & including network state in Hostinfo.

### PR DESCRIPTION
**Pull request by bradfitz**
* adds new packet "netcheck" to do the checking of UDP, IPv6, and
  nearest DERP server, and the Report type for all that (and more
  in the future, probably pulling in danderson's natprobe)
* cmd/tailscale netcheck subcommand (tentative name, likely to
  change/move) to print out the netcheck.Report.

### CODE REVIEW

File: cmd/tailscale/netcheck.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+func netcheckCmd() {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	report, err := netcheck.GetReport(ctx, log.Printf)
+	if err != nil {
+		log.Fatal(err)
+	}
+	fmt.Printf("\nReport:\n")
+	fmt.Printf("  * UDP: %v\n", report.UDP)
```
Comment (crawshaw) [inline cmd/tailscale/netcheck.go]: You don't like tabs for cli indent?

(I don't have any useful opinions here.)

File: tailcfg/tailcfg.go [hunk 2]
```diff
@@ -223,6 +224,12 @@ type Hostinfo struct {
 	RoutableIPs   []wgcfg.CIDR `json:",omitempty"` // set of IP ranges this client can route
 	Services      []Service    `json:",omitempty"` // services advertised by this machine
 
+	// Network conditions:
+	MappingVariesByDestIP opt.Bool
+	WorkingIPv6           opt.Bool
+	WorkingUDP            opt.Bool
+	DERPLatency           map[string]int // keyed by hostname, in milliseconds
```
Comment (crawshaw) [inline tailcfg/tailcfg.go]: You will need to adjust the Copy method for this.

(I've been thinking of writing a reflect-based Aliases test that walks arbitrary data structures and looks for shared pointers. But generating the test data automatically is a ton of work.)

File: tailcfg/tailcfg.go [hunk 3]
```diff
@@ -223,6 +224,12 @@ type Hostinfo struct {
 	RoutableIPs   []wgcfg.CIDR `json:",omitempty"` // set of IP ranges this client can route
 	Services      []Service    `json:",omitempty"` // services advertised by this machine
 
+	// Network conditions:
+	MappingVariesByDestIP opt.Bool
+	WorkingIPv6           opt.Bool
+	WorkingUDP            opt.Bool
```
Comment (crawshaw) [inline tailcfg/tailcfg.go]: Some forward-looking thoughts (even though we get a chance to reinvent this data structure I guess when we do control protocol v2):

- some networks block UDP entirely.
- some allow port 53
- some allow port 443 (potentially useful when people put tailscale on a machine with a real IP: we could serve out of 443 if they're not using it and people on nasty networks could skip DERP)

File: tailcfg/tailcfg.go [hunk 4]
```diff
@@ -223,6 +224,12 @@ type Hostinfo struct {
 	RoutableIPs   []wgcfg.CIDR `json:",omitempty"` // set of IP ranges this client can route
 	Services      []Service    `json:",omitempty"` // services advertised by this machine
 
+	// Network conditions:
```
Comment (crawshaw) [inline tailcfg/tailcfg.go]: I'd prefer a struct for all the network-related fields.

File: tailcfg/tailcfg.go [hunk 5]
```diff
@@ -223,6 +224,12 @@ type Hostinfo struct {
...
+	MappingVariesByDestIP opt.Bool
+	WorkingIPv6           opt.Bool
+	WorkingUDP            opt.Bool
+	DERPLatency           map[string]int // keyed by hostname, in milliseconds
+
```
Comment (crawshaw) [inline tailcfg/tailcfg.go]: A single number is probably not going to capture our latency analysis long-term. (We could use call-response keep-alives in DERP, and use those to measure RTT.) Maybe make the map value a struct and call this field "Average"?

File: netcheck/netcheck.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+	"tailscale.com/stunner"
+	"tailscale.com/types/logger"
+	"tailscale.com/types/opt"
+)
+
+type Report struct {
+	UDP                   bool                     // UDP works
+	IPv6                  bool                     // IPv6 works
+	MappingVariesByDestIP opt.Bool                 // for IPv4
+	DERPLatency           map[string]time.Duration // keyed by STUN host:port
```
Comment (crawshaw) [inline netcheck/netcheck.go]: You're using a time.Duration here, but an int in tailcfg?

File: types/opt/bool.go [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+// Package opt defines optional types.
+package opt
+
+import "strconv"
+
+// Bool represents an optional boolean to be JSON-encoded.
+// The string can be empty (for unknown or unspecified), or
+// "true" or "false".
+type Bool string
```
Comment (crawshaw) [inline types/opt/bool.go]: I'm curious if this is compatible with Swift's JSON decoder. By default, an empty opt.Bool in a struct is going to be encoded by Go as `{"B":""}`. Can Swift decode that with a `: Bool` or a `: Bool`?

I see why you're doing this (the type itself cannot carry the `omitempty` attribute) but it's unfortunate we have to use an empty string to omit a value in JSON, when JSON has a nicer way of representing this (use real boolean types and omit the field). I'd prefer a real ```MyValue bool `json:"my_value,omitempty"` ``` if we can.

File: cmd/tailscale/netcheck.go (hunk 1 above)
Comment (bradfitz) [reply cmd/tailscale/netcheck.go]: Done.

File: types/opt/bool.go (hunk 7 above)
Comment (bradfitz) [reply types/opt/bool.go]: Fixed in new separate PR.

File: netcheck/netcheck.go (hunk 6 above)
Comment (bradfitz) [reply netcheck/netcheck.go]: Discussed on chat. Difference is in-memory vs JSONable types.

File: tailcfg/tailcfg.go (hunk 5 above)
Comment (bradfitz) [reply tailcfg/tailcfg.go]: documented as "best recent time"

File: tailcfg/tailcfg.go (hunk 4 above)
Comment (bradfitz) [reply tailcfg/tailcfg.go]: Done.

File: tailcfg/tailcfg.go (hunk 3 above)
Comment (bradfitz) [reply tailcfg/tailcfg.go]: Ack

File: tailcfg/tailcfg.go (hunk 2 above)
Comment (bradfitz) [reply tailcfg/tailcfg.go]: Done.

File: go.mod [hunk 8]
```diff
@@ -20,6 +20,7 @@ require (
 	github.com/tailscale/wireguard-go v0.0.0-20200225215529-3ec48fad1002
 	golang.org/x/crypto v0.0.0-20200210222208-86ce3cb69678
 	golang.org/x/oauth2 v0.0.0-20200107190931-bf48bf16ab8d
+	golang.org/x/sync v0.0.0-20190911185100-cd5d95a43a6e
```
Comment (crawshaw) [inline go.mod]: where did this come from?

Comment (bradfitz) [reply go.mod]:
```
netcheck/netcheck.go:   "golang.org/x/sync/errgroup"
```

I'll be using that more often too.

**Pull request merged**